[ty] Check overloaded function implementation consistency

LaBatata101 opened the pull request:
<!--
Thank you for contributing to Ruff! To help us out with reviewing, please consider the following:

- Does this pull request include a summary of the change? (See below.)
- Does this pull request include a descriptive title?
- Does this pull request include references to any relevant issues?
-->

## Summary
I've separated the implementation of `is_assignable_to` into `is_parameters_assignable_to` and `is_return_type_assignable_to` as discussed in Discord, that solved the problem I was having.  But, there are still a few issues that I need some help with. 

The first one is, some tests are failing in `dataclass_transform.md`, specifically when the function is annotated with `dataclass_transform` like this one:
```python
@overload
@dataclass_transform()
def versioned_class(
    cls: T,
    *,
    version: int = 1,
) -> T: ...
@overload
def versioned_class(
    *,
    version: int = 1,
) -> Callable[[T], T]: ...
def versioned_class(
    cls: T | None = None,
    *,
    version: int = 1,
) -> T | Callable[[T], T]:
    raise NotImplementedError
```
I guess this is related to this part of the spec:

> When a type checker checks the implementation for consistency with overloads, it should first apply any transforms that change the effective type of the implementation including the presence of a yield statement in the implementation body, the use of async def, and the presence of additional decorators.
>
> https://typing.python.org/en/latest/spec/overload.html#implementation-consistency

Not sure how these transforms should be handled.

~~The second issue I have is that in this python code:~~
```python
@overload
def x(y: None = ...) -> None: ...
@overload
def x(y: int) -> str: ...


def x(y: int | None = None): ...
```
~~My implementation is not creating the diagnostic for the function implementation when it's missing the return type.~~

I'll add more tests after I solve these issues.

Closes astral-sh/ty#109
<!-- What's the purpose of the change? What does it do, and why? -->

## Test Plan

Markddown tests.
<!-- How was it tested? -->

carljm commented:
The second "issue" looks like correct behavior to me. A missing annotation is equivalent to an annotation of `Unknown/Any`, which is assignable to any type, and any type is assignable to it. So there should not be any diagnostic emitted on that code. [Mypy](https://mypy-play.net/?mypy=latest&python=3.12&gist=a76906b41281d1ab44728a5f396f0d75) does not emit any diagnostic; [pyright](https://pyright-play.net/?pyrightVersion=1.1.352&strict=true&enableExperimentalFeatures=true&code=GYJw9gtgBALgngBwJYDsDmUkQWEMpgBuApiADZgCGAJgFC0ACRpFNt1xwUAHgBRwAuKADkwKYlAC8UAHRyAlFAC0APhFjiQuTMbNyVOhy59BmFDEWqoAZxggtc%2Bu049%2BQ1PgA%2B68VJ-F5Bx0gA) does emit one, but it doesn't make any sense to me; it looks like a bug.

LaBatata101 commented:
> The second "issue" looks like correct behavior to me. A missing annotation is equivalent to an annotation of `Unknown/Any`, which is assignable to any type, and any type is assignable to it. So there should not be any diagnostic emitted on that code. [Mypy](https://mypy-play.net/?mypy=latest&python=3.12&gist=a76906b41281d1ab44728a5f396f0d75) does not emit any diagnostic; [pyright](https://pyright-play.net/?pyrightVersion=1.1.352&strict=true&enableExperimentalFeatures=true&code=GYJw9gtgBALgngBwJYDsDmUkQWEMpgBuApiADZgCGAJgFC0ACRpFNt1xwUAHgBRwAuKADkwKYlAC8UAHRyAlFAC0APhFjiQuTMbNyVOhy59BmFDEWqoAZxggtc%2Bu049%2BQ1PgA%2B68VJ-F5Bx0gA) does emit one, but it doesn't make any sense to me; it looks like a bug.

Ah okay, I was checking with `pyright`  forgot to mention that.

MichaReiser commented:
@LaBatata101 I went ahead and rebased your PR passed the Red Knot renaming change. Make sure you pull (force) before trying to push new changes.

LaBatata101 commented:
> @LaBatata101 I went ahead and rebased your PR passed the Red Knot renaming change. Make sure you pull (force) before trying to push new changes.

Thanks!

LaBatata101 commented:
@carljm could it be that tests failing for `dataclass_transform` are related to #17541 not being implemented yet?

carljm commented:
> could it be that tests failing for `dataclass_transform` are related to #17541 not being implemented yet?

I don't think so. I don't think the use of `@dataclass_transform` is even related to why these tests are failing. It looks to me like the signatures used in these tests are triggering a possible bug in our signature assignability checking. I added this debug code:

```diff
diff --git a/crates/ty_python_semantic/src/types/infer.rs b/crates/ty_python_semantic/src/types/infer.rs
index 543d8ea060..8fc98276c9 100644
--- a/crates/ty_python_semantic/src/types/infer.rs
+++ b/crates/ty_python_semantic/src/types/infer.rs
@@ -1251,6 +1251,13 @@ impl<'db> TypeInferenceBuilder<'db> {
         for (idx, overloaded_signature) in overloaded_signatures
             .iter()
             .filter(|overloaded_signature| {
+                eprintln!(
+                    "{} -- {}",
+                    impl_signature.display(self.db()),
+                    overloaded_signature.display(self.db())
+                );
+                dbg!(impl_signature.is_parameters_assignable_to(self.db(), overloaded_signature));
+                dbg!(overloaded_signature.is_return_type_assignable_to(self.db(), impl_signature));
                 !(impl_signature.is_parameters_assignable_to(self.db(), overloaded_signature)
                     && overloaded_signature.is_return_type_assignable_to(self.db(), impl_signature))
             })
```

And got this output on a failing test:

```
(cls: T | None = None, *, version: int = Literal[1]) -> T | ((T, /) -> T) -- (cls: T, *, version: int = Literal[1]) -> T
[crates/ty_python_semantic/src/types/infer.rs:1259:17] impl_signature.is_parameters_assignable_to(self.db(), overloaded_signature) = false
[crates/ty_python_semantic/src/types/infer.rs:1260:17] overloaded_signature.is_return_type_assignable_to(self.db(), impl_signature) = true
(cls: T | None = None, *, version: int = Literal[1]) -> T | ((T, /) -> T) -- (*, version: int = Literal[1]) -> (T, /) -> T
[crates/ty_python_semantic/src/types/infer.rs:1259:17] impl_signature.is_parameters_assignable_to(self.db(), overloaded_signature) = true
[crates/ty_python_semantic/src/types/infer.rs:1260:17] overloaded_signature.is_return_type_assignable_to(self.db(), impl_signature) = true
```

What this shows is that we are getting accurate signatures off of all the callables involved (and once we have a `Signature`, it's just like any other `Signature` -- there's no trace of `@dataclass_transform` left), but for some reason we do not consider the parameters `(cls: T | None = None, *, version: int = Literal[1])` assignable to the parameters `(cls: T, *, version: int = Literal[1])`. Which looks wrong to me, unless I'm missing something.

And I can reproduce this behavior without involving `dataclass_transform`: https://types.ruff.rs/c1777070-e36c-4859-9796-6f32d34fe056

So I think the next step here is to debug that minimal test case and fix it in `is_assignable_to_impl` (maybe even in a separate PR).

dhruvmanila commented:
I don't think the issue is in the logic of `is_assignable_to_impl` because using concrete types in place of the generic `T` seems to work correctly (https://types.ruff.rs/a10c8ac9-76a6-447d-965f-af23d1fa468a). There must be somewhere where the specialization is not taking place? The `internal_signature` method that's used to get the `Signature` for each of the overloads and implementation does seem to apply the specializations though.

carljm commented:
I don't think there should be any specialization happening here? These functions are unspecialized generics (as seen by the fact that the signatures printed out in my debugging above still have `T` in them), and they should be compared for assignability as unspecialized generics; they would only be specialized in the context of a call.

You're right that it looks like the issue is specific typevars; it also looks like it's specific to a bound typevar: https://play.ty.dev/7bca2116-a4f4-4e07-a518-b1077403aec4

And it looks like it's specific to comparing `T | None` vs `T`: https://play.ty.dev/1c1e1872-163b-4661-b2fd-cea0434e7e61

Ok, I just figured out what I think it is. Pushed the fix in https://github.com/astral-sh/ruff/pull/17901

Rebasing this PR on top of that should fix the dataclass-transform tests, I think.

LaBatata101 commented:
So, I've pulled the new changes made in #17910, but this case still failing
```python
@overload
def func() -> None: ...
@overload
def func[T](x: T) -> T: ...
def func[T](x: T | None = None) -> T | None:
    return x
```
I've printed the output of `is_parameters_assignable_to` and it's returning `false` for these signatures: `x: T`, `x: T | None = None`. Wasn't #17910 supposed to fix that issue? Or did I misunderstand?

https://play.ty.dev/0fc474f1-7189-4a71-92ad-5f0cd6b5e123

carljm commented:
Ugh, sorry, this is https://github.com/astral-sh/ty/issues/95 -- it impacts the PEP 695 version but not the legacy-typevar version, because in this case we consider them different typevars.

I'm a little hesitant to land this overload consistency check without fixing that bug first, since it will cause more false positives than we currently do. Sorry... any chance you're interested in trying to tackle that bug?

LaBatata101 commented:
> Ugh, sorry, this is [astral-sh/ty#95](https://github.com/astral-sh/ty/issues/95) -- it impacts the PEP 695 version but not the legacy-typevar version, because in this case we consider them different typevars.
> 
> I'm a little hesitant to land this overload consistency check without fixing that bug first, since it will cause more false positives than we currently do. Sorry... any chance you're interested in trying to tackle that bug?

I could try, but I don't understand much about type theory or the python type system, do you have any resources that I can take a look at?

carljm commented:
Probably the most useful written resource I can recommend (and this isn't saying a great deal) is the typing spec, particularly starting with the "core concepts" section. But that issue may not be the simplest place to start. Hopefully one of us can get to it before long, because I would like to land this PR -- your work here looks great.

LaBatata101 commented:
> Thanks for your contribution. I'll close this as it seems to have gone stale. Anyone interested in picking this up. Feel free to open a new PR which addresses the feedback.

This PR is just waiting for https://github.com/astral-sh/ruff/pull/18634 to get merged. I've been a little busy the last couple of days, but I'll start working on https://github.com/astral-sh/ruff/pull/18634 again soon.
